modules/ROOT/pages/xdcr-active-active-eventing.adoc

```diff
@@ -0,0 +1,155 @@
+:idprefix:
+:idseparator: -
+
+include::partial$_set_page_context.adoc[]
+
+// BEGIN -- Page Attributes
+:param-topic-group: compatibility
+:param-related: {compatibility--xref}
+:param-abstract: Learn how to design Couchbase Eventing functions to avoid replication loops and conflicts in active-active XDCR environments with Sync Gateway 4.0.
+include::partial$_show_page_header_block.adoc[]
+// END -- Page Heading
+
+== Introduction
+
+With {sgw} 4.0, you can use {server-xdcr--xref} (XDCR) alongside Eventing functions to process or enrich documents.
+
+However, in active-active XDCR environments, improper function design can create "replication" loops and data "conflicts" that compromise cluster stability.
+
+== Overview
+
+Active-active XDCR replicates documents bidirectionally between clusters.
+Eventing functions deployed on both ends of this replication topology must avoid reprocessing updated documents in a way that causes XDCR to ping-pong and never stop replicating.
+
+The following issues can arise:
+
+* You must design Eventing functions that properly modify documents in replicated buckets.
+* Without protection logic, these functions can create infinite replication loops.
+* This typically occurs when you deploy the same Eventing function on both the source and target clusters.
+
+== Understanding the Replication Loop Issue
+
+[#lbl-eventing-loop-example]
+=== Example: Enriching IP Address Documents
+
+Consider this example function, based on `case_2_enrich_ips` from the Couchbase Eventing documentation:
+
+[source,javascript]
+----
+function OnUpdate(doc, meta) {
+    log('document seen', doc);
+
+    doc["ip_num_start"] = get_numip_first_3_octets(doc["ip_start"]);
+    doc["ip_num_end"] = get_numip_first_3_octets(doc["ip_end"]);
+    log('document updated to', doc);
+    src[meta.id] = doc;
+}
+
+function get_numip_first_3_octets(ip) {
+    var return_val = 0;
+    if (ip) {
+        var parts = ip.split('.');
+        return_val = (parts[0] * (256 * 256 * 256)) + (parts[1] * (256 * 256)) + (parts[2] * 256) + parseInt(parts[3]);
+        return return_val;
+    }
+}
+----
+
+When deployed on both clusters, this function continuously reprocesses the same document:
+
+.Cluster Logs – Cluster 1
+[source,console]
+----
+... "document seen" {"ip_start":"5.62.60.1", ...}
+... "document updated to" {"ip_num_start":87964673, ...}
+... "document seen" {"ip_num_start":87964673, ...}
+----
+
+.Cluster Logs – Cluster 2
+[source,console]
+----
+... "document seen" {"ip_start":"5.62.60.1", ...}
+... "document updated to" {"ip_num_start":87964673, ...}
+----
+
+This behavior repeats indefinitely due to the mutual triggering of updates.
+
+== Preventing Redundant Updates
+
+[#lbl-prevent-redundant]
+=== Basic Protection: Field Presence Check
+
+A simple way to avoid loops is to check whether the document has already been enriched.
+
```

Comment:
For example, in the below OnUpdate Eventing function, a check has been added at the top to return if the additional fields already exist in the document.